### DISCUSSION TITLE
Bar transparency - questions on colour depth

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#2508] [Ideas] [answered]
EDIT: see PR #2513 

There's been a recurring "issue" where users have requested transparent bars (as opposed to changing the opacity which makes the widget text etc. transparent too).

I'm taking a look at this based on some old work that seems to have been done a number of years ago.

I would, however, like some guidance from the devs here (particularly @m-col given their work on Wayland) as I have some questions.

For transparency, we need 32-bit windows. My main question is why the X11 creates window based on the screen's default depth (which may not be 32-bit)
https://github.com/qtile/qtile/blob/6799b6ae5ea09454e5d4ebeb82c0f0909af39439/libqtile/backend/x11/drawer.py#L169-L173

but Wayland seems to default to a 32-bit surface:
https://github.com/qtile/qtile/blob/6799b6ae5ea09454e5d4ebeb82c0f0909af39439/libqtile/backend/wayland/drawer.py#L33

I am completely new to display servers so I'm guessing there's something different about how X11 and Wayland handle this but is there any reason why we can't have the X11 backend creating 32-bit drawers by default?

**Comment by m-col** [comment]
Good question -- do we need to use the default depth, and could we instead find a depth in `self.qtile.core.conn.default_screen.allowed_depths` that has an alpha channel? I would imagine that pretty much everybody using Qtile has a system that supports this and so we'll probably find one, and we always have the root_visual to fall back to. Perhaps there is a preferred depth that we should try to use, but check that it is supported.

The fact that the wayland draweruses that format and doesn't check that it is definitely supported I guess was an oversight, as I had just assumed it would work on most systems and that we'd get reports from users where it doesn't work (and we can adjust it accordingly). A similar approach for X might work.

It might be worth looking at some other software to see how they do it. IIRC polybar can do transparency on some colours while keeping others opaque.

**Reply by elParaguayo** [reply]
Thanks. 

I do wonder whether, if we default to 32-bit but the user's system doesn't support it, we just lose the alpha channel or does something else break?

One issue I can see here is that, for X11 (don't know about Wayland), I understand you should use pre-multiplied colours (i.e. the RGB values are adjusted by the transparency value - see [here](https://stackoverflow.com/a/39927205) for example). That would mean that we need to be aware of the user's system's ability to support this depth and adjust rgb values accordingly.

**Reply by m-col** [reply]
I'm not sure, I would have thought that cairo(cffI) would handle these kinds of things for us. By the looks of it, we are only specifying the depth on the 'final' surface i.e. XCBSurface, and all the operations on the RecordingSurface don't seem to care. Perhaps if the XCBSurface doesn't support alpha channels then cairo will just not apply them... not sure how that conversion would go (this may be a non-issue... does anyone not support alpha? idk). I'd imagine cairo would also deal with the premultipling too as it's the one doing the talking to the X server for us.

**Reply by elParaguayo** [reply]
One way to find out, I guess: I'll do some experimenting!

Separate question: if I want to test on Wayland, do I install `weston` (on Arch) and just run `qtile start -b wayland` from a TTY?

**Reply by elParaguayo** [reply]
My preference would certainly be to just create a 32-bit XCBSurface and have cairo handle the rest. I'm just not familiar with these `visuals` that get passed to the surface as they seem to come from the screen rather than being able to define a constant (like the Wayland `cairocffi.FORMAT_ARGB32`).

**Reply by m-col** [reply]
No weston involved. You need to install `wlroots`, and you'll also need `pywayland`, `pywlroots` and `python-xkbcommon`. Unfortunately those three aren't (yet) in the Arch repos, but they're all pip-able.

**Reply by elParaguayo** [reply]
Are you able to explain what visual types _are_? I've just been checking on my system and seeing what `allowed_depths` are available. I've got both 24 bit and 32 bit both of them have multiple visual types and I'm not sure what's to stop me picking any one of them!

**Reply by m-col** [reply]
I think they are classes that define things like bit depth, colour order, read/write-ability: https://tronche.com/gui/x/xlib/window/visual-types.html

Is the desired behaviour that `Drawer` copies pixels to the window and those are the final pixels, or do we want those pixels to be multipled with the pixels already there? I.e. do we set widget backgrounds to #00000000 to have their background always match that of the bar, or do we have to set the widget's background to the same hex colour as the bar's background for them to match?

**Reply by m-col** [reply]
Maybe some useful things to read in polybar's source? https://github.com/polybar/polybar/blob/master/src/x11/connection.cpp#L142

**Reply by elParaguayo** [reply]
Someone did a lot of work years ago so I've been using this as a base: https://github.com/frostidaho/qtile/commit/a7202a9d2a048296d11a9bfb455571a0c501c159

I've got a transparent bar (good) but it's not clearing properly so, when widgets redraw themselves, they overlay (bad). Something to do with these lines:
https://github.com/frostidaho/qtile/commit/a7202a9d2a048296d11a9bfb455571a0c501c159#diff-fea62743b1ec800f9f6bfe239c5e9dd7994a0a33486bdd7bae367ac96a0e7832R359-R362

**Reply by elParaguayo** [reply]
It's very weird: if I have full transparency, the widget does not clear the background so the text overwrites itself. However, if I increase the opacity the widget will, at some point, begin to redraw correctly.

**Reply by elParaguayo** [reply]
Ok. Think I know what's happening: the `clear` method is clearing the widget's `drawer` but, when this is painted to the window (which hasn't been cleared), it overlays on the previous content and so the transparency doesn't remove the previous content.

**Reply by m-col** [reply]
Really? That surprises me. I thought `CopyArea` would replace the pixels in the target with the new pixels, rather than composite them together. Perhaps we could use this to our advantage? It would mean that a widget's background can be empty pixels and it will take whatever is behind it. Of course we'll need some solution for clearing text etc.

**Reply by elParaguayo** [reply]
Well, I have been known to be wrong before!

If that's how CopyArea works then it must be clearing that's not removing the previous contents completely. More digging.

It would be nice if widgets could take the bar's background e.g. if the bar has a horizontal gradient (I've got a separate branch with gradient fills) then this is drawn in each widget rather than for the bar as a whole. Would be nice if this stretched the whole bar and widgets drew over it.

**Reply by elParaguayo** [reply]
It's not `CopyArea`.

If I dump the widget's `drawer` to a PNG file _before_ `self.drawer.draw(...` then I can still see the current contents overlaying the previous one. So it looks like `self.drawer.clear` is having problems with the transparency.

**Comment by elParaguayo** [comment, marked as answer]
Starting a new chain as the old one was getting too long.

I am pretty sure that I have now identified the problem (could paste this as an issue): Essentially the clear operator is not honoured when the recording surface writes to the XCB surface.

I wrote a short script to test the theory (I know I didn't need a class here...):
``` python
import cairocffi

class CairoTest:

    def __init__(self):
        self.reset()

    def reset(self):
        self.img_surf = cairocffi.ImageSurface(cairocffi.FORMAT_ARGB32, 200, 200)
        self.rec_surf = cairocffi.RecordingSurface(cairocffi.CONTENT_COLOR_ALPHA, (0,0,200,200))
        self.rec_ctx = cairocffi.Context(self.rec_surf)

    def save(self, fname):
        ctx = cairocffi.Context(self.img_surf)
        ctx.set_source_surface(self.rec_surf, 0, 0)
        ctx.paint()
        self.img_surf.write_to_png(f"ImageSurface{fname}.png")
        self.rec_surf.write_to_png(f"RecordingSurface{fname}.png")

test = CairoTest()

test.rec_ctx.set_source_rgba(1, 0, 0, 0.5)
test.rec_ctx.rectangle(0, 0, 100, 100)
test.rec_ctx.fill()

test.save(1)

test.rec_ctx.set_operator(cairocffi.OPERATOR_CLEAR)
test.rec_ctx.paint()
test.rec_ctx.set_operator(cairocffi.OPERATOR_OVER)

test.rec_ctx.set_source_rgba(0, 1, 0, 0.2)
test.rec_ctx.rectangle(50, 50, 100, 100)
test.rec_ctx.fill()

test.save(2)
```
The images after the first rectangle are drawn are identical:
RecordingSurface:
![RecordingSurface1](https://user-images.githubusercontent.com/946265/120921483-cf0f4300-c6bb-11eb-87c9-6a2d9e914cf9.png)
ImageSurface:
![ImageSurface1](https://user-images.githubusercontent.com/946265/120921488-d5052400-c6bb-11eb-8669-7cdb3c9fe565.png)

But after the clear and drawing the second rectangle, things look different:
RecordingSurface
![RecordingSurface2](https://user-images.githubusercontent.com/946265/120921501-e3534000-c6bb-11eb-89e6-f11d19287eee.png)
ImageSurface (didn't clear the red)
![ImageSurface2](https://user-images.githubusercontent.com/946265/120921507-eb12e480-c6bb-11eb-8fe7-a3e51db5751e.png)

If I `reset` after the save (so a new `ImageSurface` is created) then the final images look like this:
RecordingSurface
![RecordingSurface2](https://user-images.githubusercontent.com/946265/120921586-4cd34e80-c6bc-11eb-86ea-fb7b3380017e.png)
ImageSurface
![ImageSurface2](https://user-images.githubusercontent.com/946265/120921591-51980280-c6bc-11eb-9cd6-7e348f7ee9ef.png)

**Reply by elParaguayo** [reply]
Had hoped this would be an easy fix... However, if I try to recreate the `XCBSurface` by finalising it at the end of each `draw` loop, I get very weird output where widgets are drawn in the wrong places.

**Comment by elParaguayo** [comment]
Success!

![20210606_212211](https://user-images.githubusercontent.com/946265/120938946-4e793280-c70d-11eb-88e3-7ec3c7a7829a.png)

Pretty hacky solution (had to write clear direct to XCBSurface rather than the recording surface...)

Will push as a draft PR so I can get some comments.

**Comment by elParaguayo** [comment]
Moving on to Wayland and finding similar issues with clearing the surface. The `Spacer` widget is not removing content when it increases in size.

Weird.

**Reply by elParaguayo** [reply]
Solved. As the spacer is fully transparent, these lines:
``` python
        if not any(not math.isclose(0.0, i) for i in self.surface.ink_extents()):
            return
```
prevents the widget from being drawn.
